cpp/inc/bond/core/detail/any.h

```diff
@@ -0,0 +1,307 @@
+            {
+                BOOST_ASSERT(TypeId<T>::value == x._id);
+                return reinterpret_cast<const T*>(x.data());
+            }
+
+            static void emplace(any& x, const T& value)
+            {
+                new (x.data()) T{ value };
+            }
+
+            static void destroy(any& x)
+            {
+                unsafe_cast(x)->~T();
+            }
+        };
+
+        template <typename T>
+        struct impl<T, false>
+        {
+            static T* unsafe_cast(any& x) BOND_NOEXCEPT
+            {
+                BOOST_ASSERT(TypeId<T>::value == x._id);
+                return *reinterpret_cast<T**>(x.data());
+            }
+
+            static const T* unsafe_cast(const any& x) BOND_NOEXCEPT
+            {
+                BOOST_ASSERT(TypeId<T>::value == x._id);
+                return *reinterpret_cast<const T* const*>(x.data());
+            }
+
+            static void emplace(any& x, const T& value)
+            {
+                *reinterpret_cast<T**>(x.data()) = new T{ value };
+            }
+
+            static void destroy(any& x)
+            {
+                delete unsafe_cast(x);
+            }
+        };
+
+        template <typename T>
+        struct table : impl<T, (sizeof(T) <= sizeof(storage))>
+        {
+            using base = impl<T, (sizeof(T) <= sizeof(storage))>;
+
+            template <typename U = T, typename boost::enable_if<std::is_copy_assignable<U> >::type* = nullptr>
+            static void assign(any& x, const T& value)
+            {
+                *base::unsafe_cast(x) = value;
+            }
+
+            template <typename U = T, typename boost::disable_if<std::is_copy_assignable<U> >::type* = nullptr>
+            static void assign(any& x, const T& value)
+            {
+                // TODO: Cache allocated buffer and reuse.
+                base::destroy(x);
+                emplace(x, value);
+            }
+
+            static void emplace(any& x, const T& value)
+            {
+                base::emplace(x, value);
+                x._id = TypeId<T>::value;  // Update the id if emplace did not throw.
+            }
+
+            static functions make() BOND_NOEXCEPT
+            {
+                return
+                {
+                    base::destroy,
+                    [](any& x, const any& other) { assign(x, *base::unsafe_cast(other)); },
+                    [](any& x, const any& other) { emplace(x, *base::unsafe_cast(other)); },
+                    [](const any& x, const any& y) { return *base::unsafe_cast(x) == *base::unsafe_cast(y); }
+                };
+            }
+        };
+    };
+
+
+    const void* data() const BOND_NOEXCEPT
+    {
+        return &_storage;
+    }
+
+    void* data() BOND_NOEXCEPT
+    {
+        return &_storage;
+    }
+
+    template <typename T>
+    void emplace(const T& value)
+    {
+        functions::template table<T>::emplace(*this, value);
```

Comment:
Sure, I'd go with `emplace_uninitialized` but also have a comment to that effect.